### PR TITLE
Discrete-time indexed policy rollout

### PR DESCRIPTION
**Pull request by thowell**
Implements a new rollout method, `Trajectory::RolloutDiscrete`

This method uses discrete-time indices to call the policy. Useful for iLQG planner. Additionally, sets default iLQG regularization to control, which leads to qualitatively better humanoid stand performance.

### CODE REVIEW

**Comment by thowell** [comment]
@nimrod-gileadi can you think of a straightforward way to combine `Trajectory::Rollout` and `Trajectory::RolloutDiscrete`? They are effectively the same. The only difference is the policy function requires continuous- (double) or discrete-time (int) inputs

**Comment by thowell** [comment]
@nimrod-gileadi Additional changes to address unresolved items from previous PR.

**Comment by nimrod-gileadi** [comment]
> @nimrod-gileadi can you think of a straightforward way to combine `Trajectory::Rollout` and `Trajectory::RolloutDiscrete`? They are effectively the same. The only difference is the policy function requires continuous- (double) or discrete-time (int) inputs

I don't know if you'd consider it straightforward but you can do it with templates.
i.e. make a templated function, RolloutInternal in `trajectory.cc`, which takes a reference to Trajectory and all the existing arguments to rollout, as well as a function pointer for a function that can convert indexes of type T to times of type double.

Here's a starting point:

```c++
#include <functional>
#include <vector>

class Trajectory {
  void Rollout(std::function<void(double)> policy);
  void RolloutDiscrete(std::function<void(int)> policy);
  std::vector<double> times;
};

template <typename T>
inline void RolloutInternal(Trajectory* trajectory,
                            std::function<void(T)> policy,
                            T (*index_to_time)(const Trajectory*, int)) {
  for (int i = 0; i < 10; i++) {
    policy(index_to_time(trajectory, i));
  }
}

void Trajectory::Rollout(std::function<void(double)> policy) {
  RolloutInternal<double>(this, std::move(policy),
                          [](const Trajectory* trajectory, int t) -> double {
                            return trajectory->times[t];
                          });
}

void Trajectory::RolloutDiscrete(std::function<void(int)> policy) {
  RolloutInternal<int>(
      this, std::move(policy),
      [](const Trajectory* trajectory, int t) -> int { return t; });
}
```

File: mjpc/planners/gradient/planner.cc [hunk 1]
```diff
@@ -163,7 +163,10 @@ void GradientPlanner::OptimizePolicy(int horizon, ThreadPool& pool) {
 
   // copy nominal policy
   policy.num_parameters = model->nu * policy.num_spline_points;
-  candidate_policy[0].CopyFrom(policy, policy.num_spline_points);
+  {
+    const std::unique_lock<std::shared_mutex> lock(mtx_);
```
Comment (nimrod-gileadi) [inline mjpc/planners/gradient/planner.cc:167]: It can be a shared_lock.

File: mjpc/planners/ilqg/planner.cc [hunk 2]
```diff
@@ -152,9 +152,12 @@ void iLQGPlanner::OptimizePolicy(int horizon, ThreadPool& pool) {
   auto nominal_start = std::chrono::steady_clock::now();
 
   // copy nominal policy
-  candidate_policy[0].CopyFrom(policy, horizon);
-  candidate_policy[0].representation = policy.representation;
-
+  {
+    const std::unique_lock<std::shared_mutex> lock(mtx_);
```
Comment (nimrod-gileadi) [inline mjpc/planners/ilqg/planner.cc:156]: shared lock?

File: mjpc/agent.cc [hunk 3]
```diff
@@ -187,9 +187,24 @@ void Agent::Plan(std::atomic<bool>& exitrequest,
 
           // counter
           count_ += 1;
+
+          if (0) {
```
[the remark this answers is not in the document]
Comment (thowell) [reply mjpc/agent.cc]: Removed.

File: mjpc/planners/ilqg/planner.cc (hunk 2 above)
Comment (thowell) [reply mjpc/planners/ilqg/planner.cc]: Done.

File: mjpc/trajectory.h [hunk 4]
```diff
@@ -46,13 +46,20 @@ class Trajectory {
...
 
-  // simulate model forward in time
+  // simulate model forward in time (continuous, i.e., mj_step1 and mj_step2)
   void Rollout(
       std::function<void(double* action, const double* state, double time)>
           policy,
       const Task* task, const mjModel* model, mjData* data, const double* state,
       double time, const double* mocap, int steps);
 
+  // simulate model forward in time (continuous, i.e., mj_step1 and mj_step2)
```
[the remark this answers is not in the document]
Comment (thowell) [reply mjpc/trajectory.h]: Changed comments.

File: mjpc/planners/ilqg/planner.cc [hunk 5]
```diff
@@ -471,34 +475,55 @@ void iLQGPlanner::Plots(mjvFigure* fig_planner, mjvFigure* fig_timer,
...
+                    &time = this->time, &mocap = this->mocap, horizon, i]() {
+      // scale improvement
+      mju_addScl(candidate_policy[i].trajectory.actions.data(),
+                  candidate_policy[i].trajectory.actions.data(),
+                  candidate_policy[i].action_improvement.data(),
+                  improvement_step[i], model->nu * horizon);
+
+      // policy
+      auto feedback_policy = [&candidate_policy = candidate_policy, model, i](
+                                  double* action, const double* state,
```
Comment (nimrod-gileadi) [inline mjpc/planners/ilqg/planner.cc:493]: Indentation looks off, and some lines are very long below. Can you auto-format?

File: mjpc/planners/ilqg/policy.cc [hunk 6]
```diff
@@ -144,8 +144,8 @@ void iLQGPolicy::Action(double* action, const double* state,
   mju_mulMatVec(action_scratch.data(), feedback_gain_scratch.data(),
                 state_scratch.data(), dim_action, dim_state_derivative);
   mju_addTo(action, action_scratch.data(), dim_action);
-
-  // Clamp controls
+
```
Comment (nimrod-gileadi) [inline mjpc/planners/ilqg/policy.cc:147]: Trailing whitespace

File: mjpc/trajectory.cc [hunk 7]
```diff
@@ -204,6 +200,119 @@ void Trajectory::Rollout(
...
+  // initial trace
+  double* tr = SensorByName(model, data, "trace");
+  if (tr) mju_copy(trace.data(), tr, 3);
+
+  for (int t = 1; t < horizon - 1; t++) {
+    // step2
+    mj_step2(model, data);
+
+    // check for step warnings
+    if (CheckWarnings(data)) {
```
Comment (nimrod-gileadi) [inline mjpc/trajectory.cc:247]: Can you make sure that the logic to update `failure` is as in the other calls to CheckWarnings, above?

Also reset `failure` to `false` at the start of the rollout.

File: mjpc/planners/ilqg/policy.cc (hunk 6 above)
Comment (thowell) [reply mjpc/planners/ilqg/policy.cc:147]: Ran the formatter on this.

**Comment by thowell** [comment]
Let's do the Rollout function merge in the next PR. I want to implement something that is easier for the researcher to understand.

**Review by nimrod-gileadi — APPROVED** [review]

**Pull request merged**